crates/agglayer-certificate-orchestrator/src/certificate_task.rs

```diff
@@ -0,0 +1,263 @@
+                .await
+                .map_err(send_err)?;
+            let mut state = state.await.map_err(recv_err)?;
+
+            // Actually certify
+            debug!("Recomputing new state for already-proven certificate");
+            let _ = self
+                .certifier_client
+                .witness_generation(&self.certificate, &mut state)
+                .await
+                .map_err(|error| {
+                    error!(%certificate_id, ?error, "Failed recomputing the new state for already-proven certificate");
+                    error
+                })?;
+            debug!("Recomputing new state completed");
+
+            // Record the certification success
+            self.network_task
+                .send(NetworkTaskMessage::CertificateProven {
+                    height,
+                    certificate_id,
+                    new_state: state,
+                })
+                .await
+                .map_err(send_err)?;
+        }
+
+        // Second, submit settlement to L1
+        let settled_certificate = if self.header.status < CertificateStatus::Candidate {
+            debug!("Starting certificate settlement");
+            let result = self
+                .settlement_client
+                .settle_certificate(certificate_id)
+                .await?;
+            self.header.status = CertificateStatus::Settled;
+            debug!("Certificate settlement completed");
+            // Note: settle_certificate currently updates the certificate status itself.
+            // It will go to Candidate upon submission, and then up to Settled when the
+            // settlement is confirmed. However, if agglayer shuts down before
+            // settlement confirmation, we can still have a certificate with
+            // a Candidate status.
+            // TODO: all the storage related to certificate should only be touched from this
+            // struct, we should refactor
+            result.1
+        }
+        // If we're not in the shortcut path of `settle_certificate`, wait for settlement
+        // TODO: once the storage is only ever touched from here, we can make this a regular `if`
+        // For now, the only way to reach this stage is to boot with an already-candidate
+        // certificate. So we know that `self.header.settlement_tx_hash` is set.
+        else if self.header.status < CertificateStatus::Settled {
+            debug!("Resuming certificate settlement");
+            let tx_hash = self.header.settlement_tx_hash.ok_or_else(|| {
+                CertificateStatusError::SettlementError(
+                    "Settled certificate header has no settlement tx hash".into(),
+                )
+            })?;
+            if !self
+                .settlement_client
+                .transaction_exists(tx_hash.0.into())
+                .await?
+            {
+                return Err(CertificateStatusError::SettlementError(
```

Comment:
We should probably refactor `SettlementError` to be self sufficient, returning more info. What if we process multiple certificates in parallel, would we know what certificate "Settlement transaction not found" happened from the logs?